### PR TITLE
Working version for all xray-l1 requests

### PR DESCRIPTION
**Pull request by samaloney**

### CODE REVIEW

File: stixcore/products/level0/science.py [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+        structure = defaultdict(list)
+        subc_nixs = packets.get('NIX00403')[0][0].__dict__.keys()
+        for substructures in packets.get('NIX00403'):
+            for substructure in substructures:
+                for name in subc_nixs:
+                    val = getattr(substructure, name)
+                    try:
+                        structure[name].extend(val)
+                    except TypeError:
+                        structure[name].append(val)
```
Comment (samaloney) [inline stixcore/products/level0/science.py:120]: For rest of the code to work as is, which isn't necessity a great idea, this needs to be done. It probably super slow and I think could be done earlier in the packet flattening?

File: stixcore/products/level0/science.py [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+        data['time'] = (DateTime(coarse=int(control["time_stamp"]), fine=0).as_float()
+                        + data['delta_time'] + data['integration_time']/2)
+        data['timedel'] = data['integration_time']
+        data['counts'] = out_counts * u.ct
+        data['counts_err'] = out_var * u.ct
+        data['control_index'] = control['index'][0]
+        data.remove_columns(['delta_time', 'integration_time'])
+
+        data = data['time', 'timedel', 'rcr', 'pixel_masks', 'detector_masks', 'num_pixel_sets',
+                    'num_energy_groups', 'triggers', 'triggers_err', 'counts', 'counts_err']
```
Comment (nicHoch) [inline stixcore/products/level0/science.py:266]: can this be combined in one call?

File: stixcore/products/levelb/binary.py [hunk 3]
```diff
@@ -135,7 +209,8 @@ def from_tm(cls, tmfile):
         packet_data = defaultdict(list)
         for binary in tmfile.get_packet_binaries():
             packet = TMPacket(binary)
-            packet_data[packet.key].append(packet)
+            if packet.key == (21, 6, 21):
+                packet_data[packet.key].append(packet)
```
Comment (nicHoch) [inline stixcore/products/levelb/binary.py:213]: what about other packets

**Review by nicHoch — APPROVED** [review]
similar to last PR overlapping issues with the decompression, calibration ...
but we will tackle that soon.

the fix 32 energies bins sounds a bit odd to me. i thought we might handle that more dynamically. What if we have visibilities. That we can not just rebin back to more energy channels.

**Comment by samaloney** [comment]
> ``` @classmethod
> def from_levelb(cls, levelb):
>   packets = super().from_levelb(levelb)
> ```
there's no `from_levelb` on the super class?

File: stixcore/products/level0/science.py [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+            elif dl_energies[-1] == np.inf:
+                rebinned_counts[..., -1] = counts[..., -1]
+                rebinned_counts_var[..., -1] = counts_var[..., -1]
+                e_ch_end -= 1
+
+            torebin = np.where((dl_energies >= 4.0) & (dl_energies <= 150.0))
+            rebinned_counts[..., 1:-1] = np.apply_along_axis(
+                rebin_proportional, -1,
+                counts[..., e_ch_start:e_ch_end].reshape(-1, e_ch_end - e_ch_start),
+                dl_energies[torebin],  sci_energies[1:-1]).reshape((*counts.shape[:-1], 30))
```
[the remark this answers is not in the document]
Comment (samaloney) [reply stixcore/products/level0/science.py:201]: Yea so the idea is that later if you are trying to concatenate observation you need to have the same energy binning but also to serialise to fits the arrays have to match. Now for the science data we should be able to do this as it should be fixed per request. Definitely up for discussion either we do it in the fits files or later on in the software?

File: stixcore/products/level0/science.py [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+        super().__init__(service_type=service_type, service_subtype=service_subtype,
+                         ssid=ssid, control=control, data=data, **kwargs)
+        self.name = 'xray-l1'
+        self.level = 'L1'
+
+    @classmethod
+    def from_levelb(cls, levelb):
+        packets = [Packet(unhexlify(d)) for d in levelb.data['data']]
+        packets = PacketSequence(packets)
+
```
[the remark this answers is not in the document]
Comment (samaloney) [reply stixcore/products/level0/science.py:96]: I don't think there is  a `from_levelb` on the superclass

[the remark this answers is not in the document]
Comment (nicHoch) [reply stixcore/products/level0/science.py:98]: not jet

**Pull request merged**